rfd/0090-db-mfa-sessions.md

```diff
@@ -0,0 +1,176 @@
+---
+author: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+ 
+# RFD 90 - Database MFA Sessions
+
+## Required Approvers
+- Engineering: Marek `@smallinsky`, Roman `@r0mant`
+- Product: Sasha `@klizhentas`, Xin `@xinding33`
+
+## What
+Use a local proxy tunnel to define the lifetime of a database MFA session, so that database access users are not required to tap an MFA device every minute when per-session-MFA is enabled.
+
+## Why
+Database access via GUI client such as DataGrip requires the user to start a local proxy to connect to: `tsh proxy db --tunnel <db>`.
+
+To support per-session-MFA for the local proxy, we are adding a callback function called on each new connection to the local proxy.
+This callback function will check if the database certs need to be reissued to proxy the connection, and prompt for credentials/MFA if needed.
+
+Currently, database certs are issued with 1 minute ttl if per-session-MFA is enabled as described in [RFD 14](https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md).
+This causes poor UX when the GUI client can establish a new connection at any time,
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
+
+## Details
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
+should be kept in-memory by a local proxy tunnel.
+
+"Doesn't this just disable per-session-mfa for database access?" (My initial thinking)
+
+- No, not quite. Sessions are still limited by the lifetime of a local proxy process and an MFA prompt is always required to start these proxies when per-session-mfa is enabled.
+- This can "weaken" per-session-mfa security though. If a user finds that they would prefer the old behavior of 1-minute TTL certs, they can still achieve that through configuration using `max_session_ttl`.
+
+### API
+
+We should change the auth service to always issue database certs without capping the cert TTL to 1 minute for per-session-MFA.
```

Comment:
See my comment above. I think it should only apply to certificates used by the local proxy (which it'll keep in memory).

Reply:
I can point out here as well that the intention is to disable `tsh db login` and `tsh proxy db` (without --tunnel) when per-session-mfa is required.  So the auth service will "always" issue db certs without capping the TTL to 1 minute, but the client will also always use a proxy tunnel when per-session-mfa is required.

We could have the client send its intent along with the cert request, as in "I intend to keep these certs in memory" or "on disk", but I don't think it's worth doing if we disable non-tunnel clients in per-session-mfa.

So I'll rephrase this to:

```diff
### API

We should change the auth service to always issue database certs without capping the cert TTL to
1 minute for per-session-MFA; clients will only keep the newly issued database certs in-memory
when per-session-mfa is required.
```